code/crates/engine/src/consensus.rs

```diff
@@ -298,12 +301,20 @@ where
         state: &mut State<Ctx>,
         msg: Msg<Ctx>,
     ) -> Result<(), ActorProcessingErr> {
+        let is_restart = matches!(msg, Msg::RestartHeight(_, _));
+
         match msg {
-            Msg::StartHeight(height, validator_set) => {
-                self.tx_event.send(|| Event::StartedHeight(height));
+            Msg::StartHeight(height, validator_set) | Msg::RestartHeight(height, validator_set) => {
+                self.tx_event
+                    .send(|| Event::StartedHeight(height, is_restart));
 
-                // Fetch entries from the WAL
-                let wal_entries = self.wal_fetch(height).await?;
+                // Fetch entries from the WAL or reset the WAL if this is a restart
+                let wal_entries = if is_restart {
+                    self.wal_reset(height).await?;
+                    vec![]
```

Comment:
We should warn about potential node misbehavior. Also update the docs (adr-004 and/or tutorial?) with recommendation on when the application may want to reset height and other actions it might need to do, like clear state from previous hight(s) or restoring state if reset to higher heights.

Reply:
done in a3a7b161d98bc30270945b86e645fb6c746a8c08